src/osd/OSDMap.cc

```diff
@@ -5187,3 +5187,146 @@ int OSDMap::parse_osd_id_list(const vector<string>& ls, set<int> *out,
   }
   return 0;
 }
+
+void OSDMap::get_random_up_osds_by_failure_domain(int n,     // whoami
+                                                  int limit, // how many
+                                                  set<int> skip,
+                                                  set<int> *out) const {
+  int failure_domain = crush->get_top_failure_domain();
+  if (failure_domain < 1)
+    failure_domain = 1; // promote to host-level
+  auto loc = crush->get_full_location(n);
+  string my_host_name = "unknown";
+  auto it = loc.find(crush->get_type_name(1));
+  if (it != loc.end())
+    my_host_name = it->second;
+  string my_failure_domain_name = "unknown";
+  it = loc.find(crush->get_type_name(failure_domain));
+  if (it != loc.end())
+    my_failure_domain_name = it->second;
+
+  skip.insert(n); // skip myself
+  set<int> osds;
+  set<int> raw;
```

Comment:
not needed.. just use the below loop to build raw in place of the loop over raw?